diff --git a/core/modules/rest/lib/Drupal/rest/EventSubscriber/RouteSubscriber.php b/core/modules/rest/lib/Drupal/rest/EventSubscriber/RouteSubscriber.php
index d4711e0..35ce1d3 100644
--- a/core/modules/rest/lib/Drupal/rest/EventSubscriber/RouteSubscriber.php
+++ b/core/modules/rest/lib/Drupal/rest/EventSubscriber/RouteSubscriber.php
@@ -7,11 +7,13 @@
 namespace Drupal\rest\EventSubscriber;

 use Drupal\Core\Config\ConfigFactory;
+use Drupal\Core\Entity\EntityNG;
 use Drupal\Core\Routing\RouteBuildEvent;
 use Drupal\Core\Routing\RoutingEvents;
 use Drupal\rest\Plugin\Type\ResourcePluginManager;

 use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\Routing\Route;

 /**
  * Subscriber for REST-style routes.
@@ -51,7 +53,7 @@ public function __construct(ResourcePluginManager $manager, ConfigFactory $confi
    * @param \Drupal\Core\Routing\RouteBuildEvent $event
    *   The route building event.
    */
-  public function dynamicRoutes(RouteBuildEvent $event) {
+  public function resourceRoutes(RouteBuildEvent $event) {

     $collection = $event->getRouteCollection();

@@ -62,17 +64,71 @@ public function dynamicRoutes(RouteBuildEvent $event) {

         foreach ($plugin->routes() as $name => $route) {
           $route->setRequirement('_access_rest_csrf',  'TRUE');
-          $collection->add("rest.$name", $route);
+          $collection->add("rest.resource.$name", $route);
         }
       }
     }
   }
 }
